text/0000-layer-origin-metadata.md

```diff
@@ -0,0 +1,225 @@
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: https://github.com/buildpacks/rfcs/pull/63
+
+# Summary
+[summary]: #summary
+
+<!-- One paragraph explanation of the feature. -->
+
+Exposing layer origin metadata in `report.toml` will improve layer
+reproducibility and, in turn, build reproducibility.
+
+# Motivation
+[motivation]: #motivation
+
+- Layer generation can be influenced by the state of the file system,
+  namely by the state of the application directory, other layers, and
+  the stack's build image
+- Layers aren't reproducible, because we can't know enough about the
+  state of the file system at the instant that the buildpack build
+  executable is invoked (TODO: figure out if there's a less wordy name
+  for this)
+- Builds aren't reproducible, because their constituent layers aren't
+  reproducible
+- Layers and builds should be reproducible
+- Exposing layer origin metadata will enable us to know crucial facts
+  about the state of the file system at the time that the buildpack
+  build executable is invoked
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Terminology
+
+- **Layer origin metadata** is the missing piece of the
+  reproducibility puzzle; it includes the ID of the restored version
+  of the layer (if applicable), modification date, source version
+  info, and a list of the IDs of the layers that were present at layer
+  generation time.
+- **Layer ID** is the layer's unique ID, whether on disk or on the
+  registry.
+
+## Target persona
+
+This feature will be useful for **buildpack authors** (for reproducing
+layers) and **platform operators** (for reproducing builds).
+
+## Example
+
+Consider some buildpack `yabba` and some application `foo-api`. Let's
+also say that the `yabba` buildpack generates a cacheable build layer
+`yabba/build`.
+
+1. I do a build of the `foo-api` as of git SHA aaa000 involving the
+   `yabba` buildpack, without using any cache.
+2. I do a build of the `foo-api` as of git SHA aaa111 involving the
+   `yabba` buildpack, using the cache from the previous build.
+
+I want to account for all of the inputs that were present during the
+second build:
+
+1. the version of the `foo-api` application
+2. the version of the `yabba` buildpack
+3. the version of the stack's build image
+4. the version of the `yabba/build` layer that was restored from the
+   cache
+
+I can account for all but the last; I neither know that layer's
+identity, nor how it was created (the version of the `foo-api`, the
+version of the `yabba` buildpack, whatever layers existed at that
+time). Therefore, I cannot reproduce the second build.
+
+Now consider that I have access to each build's `report.toml`, and
+that the second build's `report.toml` looks something like this:
+
+```toml
+[image]
+tags = ["index.docker.io/image/name:latest", "index.docker.io/image/name:other-tag"]
+digest = "26cca5b0c787"
+previous-image-digest = "97e47fb9e0a6"
+
+[[layers]]
```

Comment:
I think this RFC could be use a schema for these new tables to define the fields needed to accomplish. I can make some educated guesses on the what these mean, but I think it makes sense to define them.

Reply:
Sure